src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,454 @@
+                        }
+                    }
+                }
+
+                // This may well be the first time that we've added something to this store.
+                Directory.CreateDirectory(_storePath);
+
+                string thumbprint = copy.Thumbprint;
+                bool findOpenSlot;
+
+                // The odds are low that we'd have a thumbprint colission, but check anyways.
+                string existingFilename = FindExistingFilename(copy, _storePath, out findOpenSlot);
+
+                if (existingFilename != null)
+                {
+                    bool dataExistsAlready = false;
+
+                    // If the file on disk is just a public key, but we're trying to add a private key,
+                    // we'll want to overwrite it.
+                    if (copy.HasPrivateKey)
+                    {
+                        try
+                        {
+                            using (X509Certificate2 fromFile = new X509Certificate2(existingFilename))
+                            {
+                                if (fromFile.HasPrivateKey)
+                                {
+                                    // We have a private key, the file has a private key, we're done here.
+                                    dataExistsAlready = true;
+                                }
+                            }
+                        }
+                        catch (CryptographicException)
+                        {
+                            // We can't read this file anymore, but a moment ago it was this certificate,
+                            // so go ahead and overwrite it.
+                        }
+                    }
+                    else
+                    {
+                        // If we're just a public key then the file has at least as much data as we do.
+                        dataExistsAlready = true;
+                    }
+
+                    if (dataExistsAlready)
+                    {
+                        // The file was added but our collection hasn't resynced yet.
+                        // Set _certificates to null to force a resync.
+                        _certificates = null;
+                        return;
+                    }
+                }
+
+                string destinationFilename;
+                FileMode mode = FileMode.CreateNew;
+
+                if (existingFilename != null)
+                {
+                    destinationFilename = existingFilename;
+                    mode = FileMode.Create;
+                }
+                else if (findOpenSlot)
+                {
+                    destinationFilename = FindOpenSlot(thumbprint);
+                }
+                else
+                {
+                    destinationFilename = Path.Combine(_storePath, thumbprint + PfxExtension);
+                }
+
+                using (FileStream stream = new FileStream(destinationFilename, mode))
```

Comment:
I started down the path of wrapping this in a `for { try { find_filename_and_write(); } catch (IOException e) { maybe_wrap_and_throw(); } }` construct, and was upset at the amount of unverifiable code that I was writing.

I think I want to just leave it as the IOException (though could be easily convinced) due to this reasoning:
* On Windows the exception has the `HResult` property which could be inspected to act on the failure reason.
* On Unix we won't have that, and emitting the IOException would at least allow the caller code to handle the failure case semantically.

The counter argument would be "Yeah, but they can just read `e.InnermostException()` and act accordingly", which I'm behind like 49.98%.  Partially because they wouldn't know to write that code.